pkg/parse/run.go

```diff
@@ -70,77 +70,78 @@ const (
 
 // RunOpts are the options used when calling Run
 type RunOpts struct {
-	runFunc RunFunc
-	backoff wait.Backoff
+	Run      RunFunc
+	Schedule events.ScheduleFunc
+}
+
+// RunResult encapsulates the result of a RunFunc.
+// This simply allows explicitly naming return values in a way that makes the
+// implementation easier to read.
+type RunResult struct {
+	CommitChanged bool
 }
 
 // RunFunc is the function signature of the function that starts the parse-apply-watch loop
-type RunFunc func(ctx context.Context, p Parser, trigger string, state *reconcilerState)
+type RunFunc func(ctx context.Context, p Parser, trigger string, state *reconcilerState) RunResult
 
 // Run keeps checking whether a parse-apply-watch loop is necessary and starts a loop if needed.
 func Run(ctx context.Context, p Parser, nsControllerState *namespacecontroller.State, runOpts RunOpts) {
 	opts := p.options()
-	// Use timers, not tickers.
-	// Tickers can cause memory leaks and continuous execution, when execution
-	// takes longer than the tick duration.
-	runTimer := opts.Clock.NewTimer(opts.PollingPeriod)
-	defer runTimer.Stop()
-
-	resyncTimer := opts.Clock.NewTimer(opts.ResyncPeriod)
-	defer resyncTimer.Stop()
-
-	retryTimer := opts.Clock.NewTimer(opts.RetryPeriod)
-	defer retryTimer.Stop()
-
-	statusUpdateTimer := opts.Clock.NewTimer(opts.StatusUpdatePeriod)
-	defer statusUpdateTimer.Stop()
+	state := &reconcilerState{}
+
+	events.Execute(ctx, runOpts.Schedule(ctx), func(event events.Event) events.HandleResult {
```

Comment:
This makes the Run function a lot smaller/simpler, but there's still room for improvement. We might want to pull out the inline handler func so we can test it independently. That said, the independent behavior doesn't really mean the whole thing works together as intented.

It might be better to move these event-specific behaviors into methods on the MetaScheduler or EventProducers.